File: javascript_version/Linked_lists/sum_list/sum_list.js
// sum list
// Write a function, sumList, that takes in the head of a linked list containing numbers as an argument. 
// The function should return the total sum of all values in the linked list.


class Node {
    constructor (value) {
        this.value = value;
        this.next = null;
    }
}


const sumList = (head) => {
    let sum = 0;
    let curr = head;
    while (curr !== null) {
        sum += curr.value;
        curr = curr.next;
    }
    return sum;
}

const sumListRecur = (head) => {
    let sum = 0;
    let curr = head;
    return _sumListRecur(curr);
}


const _sumListRecur = (curr) => {
    if (curr === null) { return 0; }
    return curr.value + _sumListRecur(curr.next);
}

console.log("iterative method");

const a = new Node(2);
const b = new Node(8);
const c = new Node(3);
const d = new Node(-1);
const e = new Node(7);

a.next = b;
b.next = c;
c.next = d;
d.next = e;

// 2 -> 8 -> 3 -> -1 -> 7

console.log(sumList(a)); // 19


const x = new Node(38);
const y = new Node(4);

x.next = y;

// 38 -> 4

console.log(sumList(x)); // 42

const z = new Node(100);

// 100

console.log(sumList(z)); // 100
console.log(sumList(null)); // 0



console.log("recursion method");
console.log(sumListRecur(a)); // 19
console.log(sumListRecur(x)); // 42
console.log(sumListRecur(z)); // 100
console.log(sumListRecur(null)); // 0

File: java_version/linked_lists/sum_list/sum_list.java
package linked_lists.sum_list;

// sum list
// Write a function, sumList, that takes in the head of a linked list containing numbers as an argument.
// The function should return the total sum of all values in the linked list.
class Node<T> {
    T val;
    Node<T> next;

    public Node(T val) {
        this.val = val;
        this.next = null;
    }
}

public class sum_list {
    public static int sumList(Node<Integer> head) {
        // todo
        int sum = 0;
        Node<Integer> curr = head;
        while (curr != null) {
            sum += curr.val;
            curr = curr.next;
        }

        return sum;
    }

    public static int sumListRecur(Node<Integer> head) {
        // todo
        Node<Integer> curr = head;
        return _sumListRecur(curr);
    }

    public static int _sumListRecur(Node<Integer> curr) {
        // todo
        if (curr == null) {
            return 0;
        }
        return curr.val + _sumListRecur(curr.next);
    }

    public static void main(String[] args) {
        Node<Integer> a = new Node<>(2);
        Node<Integer> b = new Node<>(8);
        Node<Integer> c = new Node<>(3);
        Node<Integer> d = new Node<>(-1);
        Node<Integer> e = new Node<>(7);
        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        System.out.println(sumList(a));
        System.out.println(sumListRecur(a));
        Node<Integer> x = new Node<>(38); // 19
        Node<Integer> y = new Node<>(4); // 19

        x.next = y;
        System.out.println(sumList(x)); // 42
        System.out.println(sumListRecur(x)); // 42
        Node<Integer> z = new Node<>(100);
        System.out.println(sumList(z)); // 100
        System.out.println(sumListRecur(z)); // 100
        System.out.println(sumList(null)); // 0
        System.out.println(sumListRecur(null)); // 0
    }

}